src/Qir/microsoft-quantum-qir-runtime-sys/build.rs

```diff
@@ -0,0 +1,79 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+use cmake::Config;
+use std::boxed::Box;
+use std::env;
+use std::error::Error;
+
+fn main() -> Result<(), Box<dyn Error>> {
+    println!("cargo:rerun-if-env-changed=TARGET");
+    println!("cargo:rerun-if-changed=build.rs");
+
+    for (key, value) in env::vars() {
+        println!("{}: {}", key, value);
+    }
+    if cfg!(target_os = "windows") {
+        let path_to_runtime_src = "..\\Runtime";
+        compile_runtime_libraries(path_to_runtime_src)?;     
+    } else {
+        let path_to_runtime_src = "../Runtime";
+        compile_runtime_libraries(path_to_runtime_src)?;
+    }
+
+    Ok(())
+}
+
+fn compile_runtime_libraries(path_to_runtime_src: &str) -> Result<(), Box<dyn Error>> {
+    let mut config = Config::new(path_to_runtime_src);
+    
+    if cfg!(target_os = "windows") {
+        config.static_crt(true);
+    }
+
+    set_compiler(&mut config)?;
+    set_profile(&mut config)?;
+
+    config.generator("Ninja");
+
+    let _ = config.build();
+    Ok(())
+}
```

Comment:
(Nit. I don't know Rust. Talking in C/C++ terminology)
I have an impression that the function always returns `Ok(())`. If true, then doesn't it make sense to return `void`?

Reply:
In Rust, expressions and functions always must return a value (by contrast with `void`, which is the absence of any value). Typically, a function with no meaningful return value will be marked as returning `()` (similar to Q#'s `Unit` or F#'s `unit`). Here, the return type is `Result<(), Box<dyn Error>>`, meaning it can either be `Ok(())` or `Err(t)` for some `t` of type `Box<dyn Error>`.

All that leads up to that `Ok(())` _is_ the Rust equivalent to a `void` return type in C++.

As for why it seems like it always returns `Ok(())`, the `?` operator at lines 34 and 35 cause early `return`s when either value is an `Error`. That's somewhat similar to exceptions, but has the benefit of being much more strongly typed, as it's not possible for a caller to ignore exceptions and as there's no use of exceptional control flow needed.